clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleInternalLoadBalancerAtomicOperation.groovy

```diff
@@ -0,0 +1,375 @@
+  private static final String BASE_PHASE = "UPSERT_INTERNAL_LOAD_BALANCER"
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  @Autowired
+  private GoogleOperationPoller googleOperationPoller
+
+  private final UpsertGoogleLoadBalancerDescription description
+
+  UpsertGoogleInternalLoadBalancerAtomicOperation() {}
+
+  UpsertGoogleInternalLoadBalancerAtomicOperation(UpsertGoogleLoadBalancerDescription description) {
+    this.description = description
+  }
+
+  /**
+   * curl -X POST -H "Content-Type: application/json" -d '[ { "upsertLoadBalancer": { "region": "us-central1", "ports": ["80"], "ipProtocol": "TCP", "credentials" : "my-account-name", "loadBalancerName" : "testlb", "backendService": {"name": "default-backend-service", "backends": [], "healthCheck": {"name": "basic-check", "port": 80, "checkIntervalSec": 1, "timeoutSec": 1, "healthyThreshold": 1, "unhealthyThreshold": 1, "healthCheckType": "TCP"}}, "network": "default", "subnet": "default-4464150e11ecaace", "loadBalancerType": "INTERNAL"}} ]' localhost:7002/gce/ops
+   *
+   * @param priorOutputs
+   * @return
+   */
+  @Override
+  Map operate(List priorOutputs) {
+    task.updateStatus BASE_PHASE, "Initializing upsert of load balancer $description.loadBalancerName " +
+      "in $description.region..."
+
+    if (!description.credentials) {
+      throw new IllegalArgumentException("Unable to resolve credentials for Google account '${description.accountName}'.")
+    }
+
+    def compute = description.credentials.compute
+    def project = description.credentials.project
+    def region = description.region
+    GoogleHealthCheck descriptionHealthCheck = description.backendService.healthCheck
+    String backendServiceName = description.backendService.name
+    String healthCheckName = descriptionHealthCheck.name
+    GoogleHealthCheck.HealthCheckType healthCheckType = descriptionHealthCheck.healthCheckType
+
+    // Set some default values that will be useful when doing comparisons.
+    description.ipProtocol = description.ipProtocol ?: Constants.DEFAULT_IP_PROTOCOL
+
+    ForwardingRule existingForwardingRule
+    BackendService existingBackendService
+    HealthCheck existingHealthCheck // Could be any one of Http, Https, Ssl, or Tcp.
+
+    // We first devise a plan by setting all of these flags.
+    boolean needToUpdateForwardingRule = false
+    boolean needToUpdateBackendService = false
+    boolean needToUpdateHealthCheck = false
+
+    // Check if there already exists a forwarding rule with the requested name.
+    existingForwardingRule = GCEUtil.queryRegionalForwardingRule(project, description.loadBalancerName, compute, task, BASE_PHASE)
+
+    if (existingForwardingRule) {
+      if (description.region != GCEUtil.getLocalName(existingForwardingRule.region)) {
+        throw new GoogleOperationException("There is already a network load balancer named " +
```

Comment:
Can't this naming collision be with several different types of load balancers, and not just 'network'?